### PR TITLE
Swiftinterface symbol lookup

### PR DESCRIPTION
**Pull request by adam-fowler**
On calling `definition` if a symbol returns it is in a file with a `.swiftinterface` extension it will use OpenInterface to create the generated interface file. It will then call sourcekitd `source.request.editor.find_usr` to find the location of the symbol in the generated file.

- `OpenInterfaceRequest` has been extended to include an optional symbol name.
- Broke out openInterface code in `SourceKitServer.definition` into a separate function: `respondWithInterface`, as it is called from multiple sites now
- If first occurrence of symbol definition occurs inside a `swiftInterface` file then use `respondWithInterface` to build generated interface and then find symbol location in generated interface.
- Use DocumentManager to store instances of generated interface files, to avoid generating them on every call to `openInterface` and also cache the line table.
- Added `_findUSR` function that calls `source.request.editor.find_usr`

### CODE REVIEW

File: Sources/SourceKitLSP/Swift/OpenInterface.swift [hunk 1]
```diff
@@ -81,4 +110,46 @@ extension SwiftLanguageServer {
...
+    }
+    let keys = self.keys
+    let skreq = SKDRequestDictionary(sourcekitd: sourcekitd)
+
+    skreq[keys.request] = requests.find_usr
+    skreq[keys.sourcefile] = uri.pseudoPath
+    skreq[keys.usr] = symbol
+
+    if let compileCommand = self.commandsByFile[uri] {
+      skreq[keys.compilerargs] = compileCommand.compilerArgs
```
Comment (adam-fowler) [inline Sources/SourceKitLSP/Swift/OpenInterface.swift:132]: Wasn't sure if this was necessary

File: Sources/SourceKitLSP/SourceKitServer.swift [hunk 2]
```diff
@@ -1354,6 +1354,34 @@ extension SourceKitServer {
...
+    _ req: Request<DefinitionRequest>,
+    moduleName: String,
+    symbol: String?,
+    languageService: ToolchainLanguageServer
+  ) {
+      var moduleName = moduleName
+      // Stdlib Swift modules are all in the "Swift" module, but their symbols return a module name `Swift.***`.
+      if moduleName.hasPrefix("Swift.") {
+        moduleName = "Swift"
+      }
```
Comment (adam-fowler) [inline Sources/SourceKitLSP/SourceKitServer.swift:1367]: Is there a better way to do this? Are there other modules that might be split up like this?

File: Sources/SourceKitLSP/SourceKitServer.swift [hunk 3]
```diff
@@ -1335,6 +1322,19 @@ extension SourceKitServer {
 
       switch extractedResult {
       case .success(let resolved):
+        // if first resolved location is in `.swiftinterface` file. Use moduleName to return 
+        // textual interface 
+        if let firstResolved = resolved.first, 
+           let moduleName = firstResolved.occurrence?.location.moduleName, 
+           firstResolved.location.uri.fileURL?.pathExtension == "swiftinterface" {
```
Comment (adam-fowler) [inline Sources/SourceKitLSP/SourceKitServer.swift:1329]: Is checking the extension of location uri enough here?

**Comment by bnbarham** [comment]
I thought I might give some extra background here. I haven't looked into the current infrastructure too much, but I'll mention what I would expect anyway from a Swift perspective and then we can discuss whether that's feasible to do or not.

The cursor info request returns the group name (which is where the module splitting is coming from) as well as whether it's "dynamic", ie. whether it could be overridden. In the dynamic case we will also provide the "receiver" USRs which are the statically known types of the member that's being called.

If the returned symbol isn't dynamic, then there's no need to do an index lookup - we can jump straight to the location provided by the result (*). In the case that location is a header we can make a choice:
1. Jump to the header
2. Do an index lookup for the definition and jump to that (if we have the definition)
3. Jump to the generated interface for that module
4. Provide the user a choice for one of these

If it is dynamic, we should always do an index lookup - looking up and down the hierarchy from the given receiver types. This is basically to avoid going off into a separate subtree, eg. if we had D: C, C: B, B: A, E: B and a call on `C.foo()` we wouldn't want to provide `E.foo()` as a result, but we would want to give `D.foo()` since it's possible that's the actual call.

(*) This result wouldn't be any more/less up to date than the index since we only index-while-building. Though if we instead had background indexing, it might be better to consider the index result all the time. We could also background build though so... 🤷The dynamic/receiver USR data is also in the index.

-----

With that said...

> Is checking the extension of location uri enough here?

Depends on what we actually want to do regarding the header case above. Another case to consider is what happens when library evolution isn't enabled, though I'm not sure we really want to to support that in general. I *believe* (but haven't checked) that we don't return a path at all in that case, but ideally we'd still jump to the generated interface. The index wouldn't contain any results from such modules anyway.

If this is just about providing the generated interface for swift modules, then this check works well enough.

> Is there a better way to do this? Are there other modules that might be split up like this?

Ideally we'd pass the group name to the generated interface request in this case. We could either get that by splitting (which is fine as long as we check that the symbol is from swift and not C-family) or from the cursor info request. Technically any library could provide this via `-group-info-path`, though it really only matters for when there's no source.

File: Sources/SourceKitLSP/SourceKitServer.swift (hunk 2 above)
Comment (ahoppen) [reply Sources/SourceKitLSP/SourceKitServer.swift:1367]: To extend on @bnbarham’s comment, you can specify the group name by adding the following to the `editor.open.interface` request.

```
key.groupname: "String",
```

File: Sources/SourceKitLSP/Swift/OpenInterface.swift [hunk 4]
```diff
@@ -14,30 +14,59 @@ import Foundation
...
+      // has interface already been generated
+      if let snapshot = self.documentManager.latestSnapshot(interfaceDocURI) {
+        self._findUSR(request: request, uri: interfaceDocURI, snapshot: snapshot, symbol: symbol) { result in
+          switch result {
+            case .success(let info):
+              request.reply(.success(InterfaceDetails(uri: interfaceDocURI, position: info.position)))
+            case .failure:
+              request.reply(.success(InterfaceDetails(uri: interfaceDocURI, position: nil)))
+          }
+        }
```
Comment (ahoppen) [inline Sources/SourceKitLSP/Swift/OpenInterface.swift:44]: I wonder whether we can share this code between the two `if` branches

File: Sources/SourceKitLSP/Swift/OpenInterface.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SourceKitLSP/Swift/OpenInterface.swift:132]: Compiler args are not used for the `find_usr` request, so there’s no need to specify them

https://github.com/apple/swift/blob/31032b9cae300878312dc21ddbd515c744553db6/tools/SourceKit/tools/sourcekitd/lib/Service/Requests.cpp#L3706-L3723

File: Sources/SourceKitLSP/Swift/OpenInterface.swift [hunk 5]
```diff
@@ -81,4 +110,46 @@ extension SwiftLanguageServer {
...
+    
+    let handle = self.sourcekitd.send(skreq, self.queue) { result in
+      switch result {
+      case .success(let dict):
+          if let offset: Int = dict[keys.offset],
+            let position = snapshot.positionOf(utf8Offset: offset) {
+              return completion(.success(FindUSRInfo(position: position)))
+          } else {
+            return completion(.success(FindUSRInfo(position: nil)))
+          }
```
Comment (ahoppen) [inline Sources/SourceKitLSP/Swift/OpenInterface.swift:143]: Should be indented by 2 spaces, I think.

Comment (adam-fowler) [reply Sources/SourceKitLSP/Swift/OpenInterface.swift:143]: Do you use one of the swift formats to format the code?

Comment (ahoppen) [reply Sources/SourceKitLSP/Swift/OpenInterface.swift]: No, SourceKit-LSP is manual 2 space indentation

File: Sources/SourceKitLSP/SourceKitServer.swift (hunk 2 above)
Comment (adam-fowler) [reply Sources/SourceKitLSP/SourceKitServer.swift]: I moved this code into the initialiser for OpenInterfaceRequest, made it slightly more generic and also set `key.groupname` when calling `editor.open.interface`. It appears to take an array of Strings so you can set ["Math", "Integers"] for sub-modules of sub-modules.

File: Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift [hunk 6]
```diff
@@ -20,20 +20,46 @@ public struct OpenInterfaceRequest: TextDocumentRequest, Hashable {
...
-    self.name = name
+    self.symbol = symbol
+    // Stdlib Swift modules are all in the "Swift" module, but their symbols return a module name `Swift.***`.
+    let splitName = name.split(separator: ".")
+    if splitName.count == 1 {
+      self.moduleName = name
+      self.groupNames = []
+    } else {
+      self.moduleName = String(splitName[0])
+      self.groupNames = [String.SubSequence](splitName.dropFirst()).map { String($0)}
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift:41]: DropFirst returns an ArraySlice not an Array so I don't your change would work.

File: Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift [hunk 7]
```diff
@@ -20,20 +20,46 @@ public struct OpenInterfaceRequest: TextDocumentRequest, Hashable {
...
+
+  public init(textDocument: TextDocumentIdentifier, name: String, symbol: String?) {
     self.textDocument = textDocument
-    self.name = name
+    self.symbol = symbol
+    // Stdlib Swift modules are all in the "Swift" module, but their symbols return a module name `Swift.***`.
+    let splitName = name.split(separator: ".")
+    if splitName.count == 1 {
+      self.moduleName = name
+      self.groupNames = []
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift:38]: Yeah there was a reason for that but then I changed the code and it probably isn't needed anymore

File: Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift:41]: Ah, I see. In that case: There’s a missing space before the `}`. Sorry for being so picky

[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift:41]: You need to setup a formatter then all of these things would be sorted automatically. 😄

File: Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift]: Removed special case

File: Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift]: Used String.init instead

File: Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift [hunk 8]
```diff
@@ -20,20 +20,46 @@ public struct OpenInterfaceRequest: TextDocumentRequest, Hashable {
...
 
   /// The module to generate an index for.
-  public var name: String
+  public var moduleName: String
 
-  public init(textDocument: TextDocumentIdentifier, name: String) {
+  /// The module group name.
+  public var groupNames: [String]
+
+  /// The symbol to search for in the generated module interface.
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift]: Renamed to symbolUSR

File: Tests/SourceKitLSPTests/SwiftInterfaceTests.swift
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:1]: Done

File: Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift]: Yes, I agree. We do it for swift-syntax now. We should also run swift-format on SourceKit-LSP

**Review by ahoppen — APPROVED** [review]
Thank you so much for this

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by adam-fowler** [comment]
@ahoppen Hopefully fixed the tests. Had to create a new test SwiftPM project for the swift interface tests as my edits of an existing one broke another test. Also added tests for symbols from multiple libs, with submodules and not.

**Review by ahoppen — APPROVED** [review]
Thanks

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@adam-fowler Could you also create a cherry-pick PR for release/5.9? I think it would be worth to include this improvement in the 5.9 release.

**Pull request merged**